### DISCUSSION TITLE
Register Route macro from a provider from a package.

### DISCUSSION THREAD

**Discussion by ntatoud** [adonisjs/core#4764] [Help] [answered]
Good evening ! 
I am just getting started with AdonisJs 6 (Awesome work, thank you very much ❤️) 

I am trying to build a package that adds an `openapi` macro to `Route` so that it can be used as follow : 

```ts
    router.get('/', () => { ... }).openapi(someObject)
```

For that I made an `openapi_provider.ts` file in my package :
```ts
import type { ApplicationService } from '@adonisjs/core/types';
import type { OpenApiMeta } from '../types/index.js';
import { Route } from '@adonisjs/http-server';

declare module '@adonisjs/core/http' {
	export interface Route {
		openapi(meta: OpenApiMeta): this;
	}
}

export default class OpenapiServiceProvider {
	constructor(protected app: ApplicationService) {}

	/**
 	 * Invoked by AdonisJS to register container bindings
 	 */
 	register() {}

 	/**
 	 * Invoked by AdonisJS to extend the framework or pre-configure objects
 	 */
 	async boot() {
 		Route.macro('openapi', function (this: Route, meta: OpenApiMeta) {
 			// handle metadata
 			return this;
 		});
 	}
 }

```

When i build my package, this produces a file under : './dist/providers/openapi_provider.ts'

and in my `package.json` exports field I added :
```
exports: {
    ...,
    "./openapi_provider": "./dist/providers/openapi_provider.js"
}
```
I then try to use this package in my adonisjs project (api starter kit) with `pnpm link`.

In my `adonisrc.ts` I added : 
```ts
providers: [
    () => import('my-package/openapi_provider')
]
```

And then in my route i try to do this : 
```ts
    router.get('/', () => { ... }).openapi(someObject) // typing is retrieved correctly from my-package
```

But when I try to launch my project (pnpm run dev OR node ace serve --watch) i get : 
![image](https://github.com/user-attachments/assets/c2e57c03-21df-445b-8d1f-843b09cc3fd9)

**When i try to directly put this provider in my adonis project, it works just fine somehow.**

I was wondering if it could be a problem with **pnpm link**, but i tried to replace `lucid` with a local version via pnpm link and it worked just fine 🤔 

Can anyone provide some help on this ?? 

Thank you very much 🙏🏼 
Have a great day !

**Comment by Julien-R44** [comment, marked as answer]
Yes, definitely an issue with `pnpm link`. You end up with two versions of `@adonisjs/core` in your project, and you’re adding the macro to the wrong instance.

What I recommend is using the `file` protocol to test your package locally. 
In the `package.json` of your test app, add something like :

```json
{
  "@ntatoud/openapi": "file:../path/to/your/package"
}
```

Then run `pnpm install` and it should work.  
A small drawback of this method is that you’ll need to run `pnpm install` in your test app each time you rebuild your package

**Comment by ntatoud** [comment]
Thank you for your fast answer ! It indeed fixed my problem 😄
